hydromt/components/spatial.py

```diff
@@ -28,27 +29,34 @@
 
 logger = getLogger(__name__)
 
-DEFAULT_REGION_FILE_PATH = "region.geojson"
 
-
-class ModelRegionComponent(ModelComponent):
+class SpatialModelComponent(ModelComponent, ABC):
     """Define the model region."""
 
+    DEFAULT_REGION_FILENAME = "region.geojson"
+    DEFAULT_HYDROGRAPHY_FILENAME = "merit_hydro"
+    DEFAULT_BASIN_INDEX_FILENAME = "merit_hydro_index"
+
     def __init__(
         self,
         model: "Model",
+        region_filename: StrPath = DEFAULT_REGION_FILENAME,
     ) -> None:
         super().__init__(model)
-        self._data: Optional[GeoDataFrame] = None
+        self.__filename: StrPath = region_filename
+        self.__data: Optional[GeoDataFrame] = None
 
-    @hydromt_step
-    def create(
+    def create_region(
```

Comment:
Discussed a bit with @hboisgon if these functions should be public or protected (with underscore). In a way this is [protected](https://stackoverflow.com/a/1020776/363224), since it can only be called from child classes. But I understand that it's part of the public API for a python package, but only for plugin developers, and not from a user's perspective.

For now we have gone for public, although my gut feeling says we should make them protected.

Reply:
I still think this method should be an external function ie used to be a workflow or a class as Sam started to work on.
It does a lot of nice things for region delineation which could actually be useful outside of a model component: 

- for CST to delineate a region before actually building a model 
- in export_data function to clip data from a data catalog ie no model needed)
- for clipping model (clip command), we also need to derive the region to clip to (which will then become the future model.region after clipping of all components)

Reply:
Happy to have more discussions about this, but I don't think that this should be the case. 

I have a couple of problems with putting it outside of the components: 
1. In the past putting this outside of the components caused a lot of problems with circular imports because the components need it, but it in turn uses information from the components. It's not impossible to fix, but it makes the imports very fragile which I don't like
2. I think it makes the logic much harder to follow, since then in that create region we have to do many checks for every possible region, whereas the components can just focus on their specific region kinds as well as create regions specific to them as needed. 
3. If this function is going to be part of our Public API then I would like to write down and agree on what we do and don't allow in there, since now I find it quite fluid. 
4. We're moving in the direction of making the components more responsible for these things, which has the added benefits that if plugins want to add their own region kind they can implement it in their components (perhaps delf3dfm might want to create their own `pipes` region?)

Reply:
For now I have moved the code to a `workflow`. A plugin could still decide to parse custom regions within custom components before called `parse_region`.

```py
PipesComponent(SpatialModelComponent):
  def create(self, region):
    if next(iter(region)) == "pipes":
      # do custom stuff
      geom = custom_pipes_region(region)
    else:
      geom = parse_region(region)
```

Reply:
That is true. If you think it works then I'm happy to go along with it.